static_analyzer/programming_language.py

```diff
@@ -34,10 +42,82 @@ def get_server_parameters(self) -> List[str]:
         return self.server_commands
 
     def is_supported_lang(self) -> bool:
-        """
-        Check if the language is supported by the static analyzer.
-        """
         return self.server_commands is not None
 
+    def __hash__(self):
+        return hash(self.lsp_server_key)
+
+    def __eq__(self, other):
+        if not isinstance(other, ProgrammingLanguage):
+            return False
+        return self.lsp_server_key == other.lsp_server_key
+
     def __str__(self):
-        return f"ProgrammingLanguage(language={self.language}, size={self.size}, percentage={self.percentage}, suffixes={self.suffixes})"
+        return f"ProgrammingLanguage(language={self.language}, lsp_server_key={self.lsp_server_key}, size={self.size}, percentage={self.percentage:.2f}%, suffixes={self.suffixes})"
+
+
+class ProgrammingLanguageBuilder:
+    """Builder to create ProgrammingLanguage instances from tokei output with greedy LSP matching."""
+
+    def __init__(self, lsp_configs: dict):
+        self.lsp_configs = lsp_configs
+        # Build reverse index: extension -> lsp_config_key
+        self._extension_to_lsp: Dict[str, str] = {}
```

Comment:
nit:
```suggestion
        self._extension_to_lsp: dict[str, str] = {}
```

I see it in a couple other places in the builder class. Thank you for the types and again sorry for the original confusion with the types (the main idea of the latest PEP is to reduce the dependencies on typing).